### ISSUE TITLE
Migrate to .NET 4.7.1 and to SharpDX

### ISSUE THREAD

**Issue by ericpignet** [limyz/DTXmaniaNX#55]
The project is currently using very old Dot Net framework 3.5, SlimDX library which was abandoned years ago, and SoundDecoder which is also abandonned.

Other DTXMania forks have already upgraded to recent .NET versions and to [SharpDX](http://sharpdx.org/) which is an active C# DirectX wrapper. Original DTXMania did the migration first, and DTXMania Ver.K did it in version 4.10, DTXMania AL did it too.
We need to do it too, in order to benefit from the optimisations brought by recent .NET framework, better quality libraries, especially for sound decoding., 
This is also a prerequisite for a future migration to 64 bit (already done by DTXMania, VerK 4.x and AL) which should optimize memory access for people with large simfile libraries.
Finally this will make it easier to backport features or fixes done in other forks.

**Comment by ericpignet** [comment]
I've been working on that for a while, trying to manually backport from Ver.K 4.10, with adaptations when needed.
See my branch: https://github.com/ericpignet/DTXmaniaXG/tree/sharpdx

I had to reorganise the directory structure to make the build process work, and to align with DTXMania and latest Ver.K.
I think it's better/clearer now, my only concern is that most files look deleted/recreated.

**Comment by limyz** [comment]
@fisyher and myself highly recommend this adaptation. As SharpDX is no longer in active development, do consider [Vortice.Windows](https://github.com/amerkoleci/Vortice.Windows) as an alternative.

**Comment by fisyher** [comment]
SharpDX is a good choice as it is very stable and used in many projects.
 
Which library should we use to replace SoundDecoder?

**Comment by tanabotadtx** [comment]
Original dtxmania seems already removed SoundDecoder... (BASS.DLL is used instead of SoundDecoder)

There are mainly 2 commits to implement it;
https://osdn.net/projects/dtxmania/scm/git/dtxmania/commits/397438f86142fd5d0777977b67b4db3bd44c7a0a
https://osdn.net/projects/dtxmania/scm/git/dtxmania/commits/8cb598283e68b1b0fc3cc2e851b5637547842866

First one is to remove SoundDecider.dll, second one is to fix degrade (for keeping legacy support)

But it seems there are many commits about audio support.... it might be better to use whole source code of the latest original ones, instead of patching to our code.

**Comment by limyz** [comment]
Our fork uses [Un4seen](http://www.un4seen.com) for audio support.

DirectSound has been superseded by [XAudio2](https://docs.microsoft.com/en-us/windows/win32/xaudio2/xaudio2-introduction). According to [Low Latency Audio in the Windows Hardware documentation](https://docs.microsoft.com/en-us/windows-hardware/drivers/audio/low-latency-audio), we can natively use Windows 10's AudioGraph or WASAPI. [ASIO should come from Steinberg](https://www.steinberg.net/en/company/developers.html).

**Comment by ericpignet** [comment]
I'm really following what was done in other DTXMania versions, not more.
SoundDecoder is replaced by:

- BASS (from Un4seen) for MP3 and Ogg. See https://github.com/ericpignet/DTXmaniaXG/blob/sharpdx/FDK/Code/03.Sound/Cmp3ogg.cs (this class is coming straight from Ver.K 4.10)
- [BJXA](https://github.com/Dridi/bjxa) for XA format. Same as DTXMania and Ver.K 4.10. What's good with this library is that it's free software, we can include the source code in the project, do changes if needed, rebuild with more recent compilers etc. Not one old Dll found somewhere on the internet.

I didn't know of Vortice.Window. Indeed that would be the best choice going forward, but this will be much more work than what I did (backporting the migration to SharpDX done by other version).
Migration to SharpDX is already a first step forward and made it possible to use recent .NET framework. Also our version of FDK becomes very close to Ver.K 4.10 so we can easily integrate fixes from them. That made it easy to backport the WASAPI Exclusive/Shared separate modes.
Let's keep Vortice in mind for the future.

Same for migration from DSound to XAudio2. I agree we should do that, but it's much more work than what I did (which was already not so easy as the commits in Ver.K were in the middle of tens of unrelated changes).

We can open another wishlist item for Vortice, and one for XAudio2 :-)
